include/onnxruntime/core/session/onnxruntime_cxx_inline.h

```diff
@@ -288,6 +288,85 @@ inline char* Session::EndProfiling(OrtAllocator* allocator) const {
+                                                   const std::vector<std::string>& output_names, const RunOptions& run_options) {
+  size_t output_node_count = GetOutputNames().size();
+  std::vector<Value> output_values;
+  for (size_t i = 0; i < output_node_count; i++) output_values.emplace_back(nullptr);
+  Run(input_names, input_values, output_names, output_values, run_options);
+  return output_values;
+}
+
+inline void ExperimentalSession::Run(const std::vector<std::string>& input_names, const std::vector<Value>& input_values,
+                                     const std::vector<std::string>& output_names, std::vector<Value>& output_values, const RunOptions& run_options) {
+  size_t input_node_count = input_names.size();
+  size_t output_node_count = output_names.size();
+  std::vector<const char*> input_names_(input_node_count, nullptr);
+  size_t i = 0;
+  for (auto it = input_names.begin(); it != input_names.end(); it++) input_names_[i++] = (*it).c_str();
+  std::vector<const char*> output_names_(output_node_count, nullptr);
+  i = 0;
+  for (auto it = output_names.begin(); it != output_names.end(); it++) output_names_[i++] = (*it).c_str();
+  Session::Run(run_options, input_names_.data(), input_values.data(), input_node_count, output_names_.data(), output_values.data(), output_node_count);
+}
+
+inline std::vector<std::string> ExperimentalSession::GetInputNames() const {
+  Ort::AllocatorWithDefaultOptions allocator;
+  size_t node_count = GetInputCount();
+  std::vector<std::string> out(node_count);
+  for (size_t i = 0; i < node_count; i++) {
+    char* tmp = GetInputName(i, allocator);
+    out[i] = tmp;
+    allocator.Free(tmp); // prevent memory leak
+  }
+  return out;
+}
+
+inline std::vector<std::string> ExperimentalSession::GetOutputNames() const {
+  Ort::AllocatorWithDefaultOptions allocator;
+  size_t node_count = GetOutputCount();
+  std::vector<std::string> out(node_count);
+  for (size_t i = 0; i < node_count; i++) {
+    char* tmp = GetOutputName(i, allocator);
+    out[i] = tmp;
+    allocator.Free(tmp); // prevent memory leak
+  }
+  return out;
+}
+
+inline std::vector<std::string> ExperimentalSession::GetOverridableInitializerNames() const {
+  Ort::AllocatorWithDefaultOptions allocator;
+  size_t init_count = GetOverridableInitializerCount();
+  std::vector<std::string> out(init_count);
+  for (size_t i = 0; i < init_count; i++) {
+    char* tmp = GetOverridableInitializerName(i, allocator);
+    out[i] = tmp;
+    allocator.Free(tmp); // prevent memory leak
+  }
+  return out;
+}
+
+inline std::vector< std::vector<int64_t> > ExperimentalSession::GetInputShapes() const {
+  size_t node_count = GetInputCount();
+  std::vector<std::vector<int64_t>> out(node_count);
+  for (size_t i = 0; i < node_count; i++) out[i] = GetInputTypeInfo(i).GetTensorTypeAndShapeInfo().GetShape();
+  return out;
```

Comment:
A shape can have symbolic (has a name in dim_param) or unknown dimensions (has neither a value or a name). Those will be represented as -1 values in this vector, but that doesn't provide a signal to the user that more work needs to be done to make the shape usable. I can imagine a user calling GetInputShapes and trying to create an input buffer based on that and getting confused when it fails if there's a dimension value of -1. 

Would it be better to take the vector as a non-const reference and return a bool indicating whether the shape has no symbolic/unknown dimensions to make this more explicit?

Or at the very least there needs to be documentation that calls out that this vector may have -1 values which must be handled appropriately.

Reply:
Added a comment about symbolic dimension values [here](https://github.com/jgbradley1/onnxruntime/blob/modern-C%2B%2B-API/include/onnxruntime/core/session/onnxruntime_cxx_api.h#L263).  I think the logic of checking for a negative value before creating an input buffer is best left up to the API caller or placed in a brand new API function. It's outside the scope of expectation for what the `Get*Shapes()` functions are supposed to do.

As for taking in a non-const reference and returning a bool...I'm not in favor of that design at this time because there can be many interpretations of a bool response....for example, a `false` return value may indicate the lack of a symbolic dimension OR just a failed call to the model API; It would not be very clear to the API user and would not offer any guarantee that something else in the session or model hasn't failed.

I would rather have a new helper API function like `HasSymbolicDimension()` (or whatever name makes sense) before messing with the `Get*Shapes()` functions.

Reply:
Your suggestion about a user trying to create an input buffer and getting confused because of a symbolic dimension - I'd like to see some kind of error checking for that use case in the [`CreateTensor()`](https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/session/onnxruntime_cxx_api.h#L256) functions. Right now, nothing gets checked and it just calls the C API.

Reply:
A separate PR has been submitted to catch errors when someone tries to create a tensor with a symbolic dimension in the shape.

https://github.com/microsoft/onnxruntime/pull/4297

Reply:
A separate PR has been submitted to catch errors when someone tries to create a tensor with a symbolic dimension in the shape.

https://github.com/microsoft/onnxruntime/pull/4297